Allow to add/remove citations in notes

mrtcode opened the pull request:
As discussed in #2284, this is an experimental PR that allows to add/remove citations that are immediately after annotation. Citations in other places or multi-item citations are untouched.

It now also shows a title with formatted citation when hovering a highlight or image annotation.

For now the default behavior (whether to show colors or citations, or not) is not changed and it also doesn't remember the last setting yet. Firstly let's try this out and make sure actually adding/removing citations is the right way to go.

It actually feels much cleaner without citations.

dstillman commented:
So I guess my worry here would be that we make it very easy to create a version of the note without the annotations cited. It makes sense as a display option, and it makes a little more sense when using "Include Zotero Links" for Markdown export, but when you use Add Note or note export without links, it's a little dangerous (though, of course, you might have inserted a citation for the item manually). Do we want export-time options to add citations back when using those methods?

Otherwise:

I think "Remove Citations" is too scary. It's a little imprecise, but "Hide Citations" (and "Show Citations") is probably better, since it conveys that it's reversible. And maybe same for colors?

And, actually, doesn't it need to be "Hide[/Show] Annotation Citations"? Because it wouldn't affect standalone citations.

Maybe put a separator above the annotation options, since there will be 2–3 of them now?

Maybe we should have "Hide Citation" show up in the popup for an individual citation? You can delete manually, but that's tedious, and it would mirror "Add Citation" (which should probably be "Show Citation") on the annotation.

dstillman commented:
The colors toggle would arguably make sense as an export-option too. So maybe both of them are both display options and export options?

mrtcode commented:
So we have three ways to export a note: 1) export as file, 2) export to clipboard, 3) "Add Note" from processor. We can add the necessary export options to 1) and 2), but for 3) we don't have any way to do that. 

Next, an export option for colors only makes sense for "Add Note" and Rich Text or HTML export.

dstillman commented:
As a temporary solution until we have a proper redesign, we could probably just add the Z menu from the Quick Format dialog to the Add Note dialog? (@adomasven)

I'm not sure what the wording on these would be. It's not really "Include Annotation Colors"/"Include Citations", because that implies that if you disable them you don't get citations or colors, but that's not the option, right? It's forcing them to be included even when they're not. "Include All Annotation Colors"/"Include All Citations"?

mrtcode commented:
Ok, we can review this now.
Various little things from other issues are fixed here as well.

1) Removed "Insert Citation" from context menu #2308.
2) No longer shows "Go to Page" when the citation immediately follows an annotation for the same page #2306.
3) Citation button in the toolbar restores the citation if cursor is inside an annotation #2307.
4) Added a separator to the three dots menu, although the word "annotation" feels to be used excessively.
5) Opens the best PDF attachment #2306.
6) Color and citation add/remove words are now renamed to show/hide.
7) Add "Hide Citation" option to citation popup.
